kore/src/Kore/Unification/SubstitutionNormalization.hs

```diff
@@ -78,130 +73,190 @@ x = f(x) or something equivalent).
-                            Right (SetCtorCycle nonSimplifiableCycle)
-                          | otherwise ->
-                            Right (MixedCtorCycle nonSimplifiableCycle)
-                        Right _ ->
-                            Left (NonCtorCircularVariableDependency vars)
-                Right result -> Right (Sorted result)
-          where
-            nonSimplifiableSortResult =
-                topologicalSort (Set.toList <$> nonSimplifiableDependencies)
-    case topologicalSortConverted of
-        Left err -> throwError err
-        Right (MixedCtorCycle _) -> return Predicate.bottom
-        Right (Sorted vars) -> lift $ normalizeSortedSubstitution' vars
-        Right (SetCtorCycle vars) -> normalizeSubstitution
-            $ Map.mapWithKey (makeRhsBottom (`elem` vars)) substitution
+normalizeSubstitution substitution =
+    maybe bottom fromNormalization $ normalize substitution
+  where
+    bottom = return Predicate.bottom
+    fromNormalization Normalization { normalized, denormalized }
+      | null denormalized =
+        pure
+        $ Predicate.fromSubstitution
+        $ Substitution.unsafeWrap normalized
+      | otherwise =
+        throwError (SimplifiableCycle variables)
+      where
+        (variables, _) = unzip denormalized
+
+{- | The result of /normalizing/ a substitution.
+
+'normalized' holds the part of the substitution was normalized successfully.
+
+'denormalized' holds the part of the substitution which was not normalized
+because it contained simplifiable cycles.
+
+ -}
+data Normalization variable =
+    Normalization
+        { normalized, denormalized :: !(UnwrappedSubstitution variable) }
+
+instance Semigroup (Normalization variable) where
+    (<>) a b =
+        Normalization
+            { normalized = Function.on (<>) normalized a b
+            , denormalized = Function.on (<>) denormalized a b
+            }
+
+instance Monoid (Normalization variable) where
+    mempty = Normalization mempty mempty
+
+{- | 'normalize' a substitution as far as possible.
+
+The substitution is given as a 'Map', so there can be no duplicates.
+
+The result is @Nothing@ if the substitution is not satisfiable, for example
+because it contains pairs such as @x = \\bottom@ or because it contains
+constructor cycles with element variables.
+
+ -}
+normalize
+    ::  forall variable
+    .   SubstitutionVariable variable
+    =>  Map (UnifiedVariable variable) (TermLike variable)
+    -- ^ De-duplicated substitution
+    ->  Maybe (Normalization variable)
+normalize (dropTrivialSubstitutions -> substitution) =
+    case topologicalSort allDependencies of
+        Right order -> sorted order
+        Left _ ->
+            case topologicalSort nonSimplifiableDependencies of
+                Right variables ->
+                    -- Removing the non-simplifiable dependencies removed the
+                    -- cycle, therefore the cycle is simplifiable.
+                    simplifiableCycle variables
```

Comment:
Does this work properly with multiple cycles (a->b, b->a, c->d, d->c)? Or with cycles one within another (a->b, b->c, c->a, c->b)? Are there tests for that?

Reply:
It works as well as it ever did, but there are not many tests for cycles. I will refresh the tests.

Reply:
I'm not sure how to reply to "works as well as it ever did".

My point was that code that worked before when a cycle was treated as "a generic error in the substitution, so we need to dump the entire thing in the predicate" might or might not be good enough if you're splitting the substitution, i.e. when you're treating the cycle as an error with some very specific details.

I.e. if you have two cycles, does the topological sort algorithm return only one of them? If yes, would the current code consider the remaining cycle as "normalized" or "denormalized"?

I don't remember how the topological sorting algorithm works, and the current code may work properly, but I thought I'd ask. I would have checked this more carefully, but I wasn't one of the assigned reviewers, so I enjoyed the lack of responsibility and just sent a thought that crossed my head when reading the code. Sorry if I bothered you for nothing.

BTW, I didn't see any (new) tests for the issues mentioned above, not sure if that's because it's too late and I missed them, or they weren't needed, or they were needed, but the current code didn't work and required too many fixes, or whatever.

Reply:
> My point was that code that worked before when a cycle was treated as "a generic error in the substitution, so we need to dump the entire thing in the predicate" might or might not be good enough if you're splitting the substitution, i.e. when you're treating the cycle as an error with some very specific details.

This doesn't change the behavior of substitution simplification: _any_ cycle is still a generic error, and we dump the entire substitution into the predicate. This is only a refactoring so that there is a function in the substitution normalizer that could possibly recover. In the original code, we throw a `Left` as soon as there is a problem, and there's practically no chance of recovering. In the refactored code, there is a wrapper that throws `Left` and an inner part which splits the substitution, but the wrapper throws `Left` if the substitution is split at all. The wrapper is the only part that is actually used during simplification, so the behavior of substitution normalization is the same as always. (That's what I meant by "it works as well as it ever did": I didn't change how this code works, it's only moving things around. I'm sorry, I realize now that it sounded snide.)

I haven't changed the substitution-splitting behavior yet because I need to separate the predicate simplifier used for unification from the one used for ordinary simplification. The unification simplifier should regard a denormalized substitution as an error instead of splitting on it, but we can't do that yet.